matrix-sdk/src/main/java/org/matrix/androidsdk/rest/client/MediaScanRestClient.java

```diff
@@ -15,47 +15,152 @@
+                public void onMatrixError(MatrixError e) {
+                    if (e.mStatus == 404) {
+                        // On 404 consider the public key is not available, so do not encrypt body
+                        mMxStore.setAntivirusServerPublicKey("");
+
+                        callback.onSuccess("");
+                    } else {
+                        super.onMatrixError(e);
+                    }
+                }
+            }));
+        }
+    }
+
+    /**
+     * Reset Antivirus server public key on cache
+     */
+    public void resetServerPublicKey() {
+        if (mMxStore != null) {
+            mMxStore.setAntivirusServerPublicKey(null);
+        }
+    }
+
     /**
      * Scan an unencrypted file.
      *
      * @param domain   the server name extracted from the matrix content uri
-     * @param mediaId   the media id extracted from the matrix content uri
+     * @param mediaId  the media id extracted from the matrix content uri
      * @param callback on success callback containing a MediaScanResult object
      */
     public void scanUnencryptedFile(final String domain, final String mediaId, final ApiCallback<MediaScanResult> callback) {
-
         mApi.scanUnencrypted(domain, mediaId).enqueue(new DefaultRetrofit2CallbackWrapper<>(callback));
     }
 
     /**
      * Scan an encrypted file.
      *
      * @param encryptedMediaScanBody the encryption information required to decrypt the content before scanning it.
-     * @param callback on success callback containing a MediaScanResult object
+     * @param callback               on success callback containing a MediaScanResult object
      */
     public void scanEncryptedFile(final EncryptedMediaScanBody encryptedMediaScanBody, final ApiCallback<MediaScanResult> callback) {
+        // Encrypt encryptedMediaScanBody if the server support it
+        getServerPublicKey(new SimpleApiCallback<String>(callback) {
+            @Override
+            public void onSuccess(String info) {
+                Call<MediaScanResult> request;
+
+                // Encrypt the data, if antivirus server supports it
+                if (!TextUtils.isEmpty(info)) {
+                    try {
+                        OlmPkEncryption olmPkEncryption = new OlmPkEncryption();
+                        olmPkEncryption.setRecipientKey(info);
+
+                        String data = JsonUtils.getCanonicalizedJsonString(encryptedMediaScanBody);
+
+                        OlmPkMessage message = olmPkEncryption.encrypt(data);
+
+                        EncryptedMediaScanEncryptedBody encryptedMediaScanEncryptedBody = new EncryptedMediaScanEncryptedBody();
+                        encryptedMediaScanEncryptedBody.encryptedBodyFileInfo = new EncryptedBodyFileInfo(message);
+
+                        request = mApi.scanEncrypted(encryptedMediaScanEncryptedBody);
+                    } catch (OlmException e) {
+                        // Fallback to unencrypted body (should not happen)
```

Comment:
If we failed here, we must not send the unencrypted body for security reason
Please return an error

Reply:
ok